### PR TITLE
Update simulators.

### PR DESCRIPTION
**Pull request by sergeisakov**
Replace generated CPU simulators with handwritten CPU simulators.

### CODE REVIEW

**Review by 95-martin-orion — COMMENTED** [review]
In lieu of reviewing all 20k lines of this, I have a couple of inline comments and some high-level questions:

1. Since the AVX512 tests don't run during CI, could you verify that they pass locally?
2. How does this change affect performance, if at all?

**Comment by sergeisakov** [comment]
1. Yes, the AVX512 tests pass locally.

2. This change doesn't affect performance for medium and large circuits. There is a small improvement in performance for small circuits (less than 10 qubits, single thread).

File: pybind_interface/Makefile [hunk 1]
```diff
@@ -11,7 +11,7 @@ QSIMLIB_DECIDE = ../qsimcirq/qsim_decide`python3-config --extension-suffix`
 PYBINDFLAGS_BASIC = -Wall -shared -std=c++17 -fPIC `python3 -m pybind11 --includes`
 PYBINDFLAGS_SSE = -msse4.1 -Wall -shared -std=c++17 -fPIC `python3 -m pybind11 --includes`
 PYBINDFLAGS_AVX2 = -mavx2 -mfma -Wall -shared -std=c++17 -fPIC `python3 -m pybind11 --includes`
-PYBINDFLAGS_AVX512 = -mavx512f -Wall -shared -std=c++17 -fPIC `python3 -m pybind11 --includes`
+PYBINDFLAGS_AVX512 = -mavx512f -mbmi2 -Wall -shared -std=c++17 -fPIC `python3 -m pybind11 --includes`
```
[the remark this answers is not in the document]
Comment (sergeisakov) [reply pybind_interface/Makefile:14]: Yes, all AVX512 devices support BMI2.

File: lib/simulator_basic.h [hunk 2]
```diff
@@ -50,22 +50,22 @@ class SimulatorBasic final {
 
     switch (qs.size()) {
     case 1:
-      ApplyGate1H(qs, matrix, state);
+      ApplyGateH<1>(qs, matrix, state);
```
[the remark this answers is not in the document]
Comment (sergeisakov) [reply lib/simulator_basic.h:53]: Perhaps it makes sense to extend controlled gates up to 6 target qubits (note that the number of control qubits is not limited). 7-qubit gates become quite slow.

File: lib/simulator.h [hunk 3]
```diff
@@ -0,0 +1,471 @@
...
+    }
+  }
+
+  struct Masks1 {
+    uint64_t imaskh;
+    uint64_t qmaskh;
+  };
+
+  template <unsigned H, unsigned R>
+  static Masks1 GetMasks1(const std::vector<unsigned>& qs) {
```
[the remark this answers is not in the document]
Comment (sergeisakov) [reply lib/simulator.h:180]: "N" is not a reference to anything. "N" just denotes the order in which these functions were added. I added a comment that describes these functions.

**Review by 95-martin-orion — APPROVED** [review]
No glaring issues stood out on a quick skim of the `simulator_*` files, so since the tests are passing I think this is good to go.

**Pull request merged**